### ISSUE TITLE
Schema change takes very long to complete on empty table

### ISSUE THREAD

**Issue by roncrdb** [cockroachdb/cockroach#42224]
User has been experiencing severe latency when running a schema change on an **empty** table. 

The schema change being run is: 

```sql 
ALTER TABLE t1 PARTITION BY LIST (locality) (
PARTITION uswest VALUES IN (‘us-west’),
PARTITION uscentral VALUES IN (‘us-central’),
PARTITION asiasoutheast VALUES IN (‘asia-southeast’)
);
```

The user reports this schema change taking over 3 minutes to complete. 

I’ve tried reproducing this issue on my end but was not able to, however I believe @andy-kimball has and he may be able to shed more light as to what is causing this in a geo-distributed cluster.

I tried reproducing this using roachprod and setting up nodes in the same regions as the user. 

I then tired running a similar schema change on the following table: 

```
show create table users;
  table_name |                      create_statement
+------------+-------------------------------------------------------------+
  users      | CREATE TABLE users (
             |     id UUID NOT NULL,
             |     city STRING NOT NULL,
             |     name STRING NULL,
             |     address STRING NULL,
             |     credit_card STRING NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (city ASC, id ASC),
             |     FAMILY "primary" (id, city, name, address, credit_card)
             | )
(1 row)
```

On an empty table the schema change took about 1.1s

```
ALTER TABLE users PARTITION BY LIST (city)
(
PARTITION uswest VALUES IN ('seatle', 'san francisco', 'los angeles'),
PARTITION uscentral VALUES IN ('oklahoma city', 'dallas', 'fort worth'),
PARTITION useast VALUES IN ('new york', 'boston', 'washington dc')
);
ALTER TABLE

Time: 1.134117579s
```

With about 1k rows, it took about 1.5 seconds, still not anywhere near the 3 minutes reported by the original user.

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[4021](https://cockroachdb.zendesk.com/agent/tickets/4021) has been linked to this issue.

**Comment by awoods187** [comment]
@andy-kimball i believe you experienced something similar

**Comment by andy-kimball** [comment]
I never experienced multi-minute delays. More like 20-30 second delays on some operations. I wonder if the cluster is loaded and that's what causing additional delays. My understanding (though the SQL/Bulk I/O teams would know best) is that the issue is that the schema change operations perform many sequential round-trips that add up. They would need to be heavily parallelized to fix this issue.

**Comment by roncrdb** [comment]
@dt @lucy-zhang any thoughts on what @andy-kimball mentioned above?

**Comment by dt** [comment]
sorry to keep passing the ball but partitioning is owned by the SQL team, not bulk-io. I don't know what it actually does under the hood -- i.e. if its is in a loop writing new zones or something. I think... @solongordon might know?

**Comment by dt** [comment]
to clarify a bit more "schema changes" is very broad -- some, like dropping a db with many tables, have been observed to have some O(n) loops while others, like renaming a table, are pretty simple, and different teams own different ones, so I don't really have a one-size-fits-all answer about how long a "schema change" would take.

**Comment by solongordon** [comment]
My understanding is that a PARTITION BY statement just involves updating an index descriptor. The runtime should be irrelevant to how much data is in the table. I don't know what would cause it to take minutes.

**Comment by jordanlewis** [comment]
@dt I thought we were discussing the other day that a common cause for multi-minute delays was waiting for lease expirations. I agree that partitioning doesn't do any work under the hood besides what @solongordon said.

**Comment by andy-kimball** [comment]
@jordanlewis, it sounds like your team is the top-level owner of partitioning, so can someone on your team dig into this further? Assuming, that person proves that it's a lower-level component causing the issues, they can post the supporting information / give lower level repro, and then pass it along to the best team.

**Comment by jordanlewis** [comment]
At the moment, to me it feels like there is insufficient information to go on here to do any further investigation. We don't have a repro, logs, a data dump, or even many details on the setup, so all we can do is speculate on what might have happened, which usually isn't a great use of time.

We do know that there are like 7-8 roundtrips per schema change caused by things like checking the comments table. This stuff is easily fixable and we can do it, but it doesn't add up to minutes like we've discussed.

@roncrdb, perhaps the user could chime in with some more information here?

**Comment by roncrdb** [comment]
Hey @jordanlewis ,

So this was on v19.2.0-rc2 I have the users complete DDL for this database and have tried to reproduce this on my end, but no luck, I'm seeing an ALTER TABLE take ~160ms which is even less than what I had initially reported when just using my own tables. 

@ jtakamine Could you give our engineers a bit more insight, I know you mentioned you no longer have this cluster and have since reverted back to using 19.1.5, but we may be able to still get some useful information from you directly. 

Thanks!

**Comment by roncrdb** [comment]
User reported this as solved in 19.2.1
